zk_toolbox/crates/zk_inception/src/utils/ports.rs

```diff
@@ -0,0 +1,288 @@
+            }
+        }
+
+        Ok(ecosystem_ports)
+    }
+
+    /// Scans the given directory for YAML files in the immediate directory only (non-recursive).
+    /// Processes each YAML file found and updates the EcosystemPorts accordingly.
+    fn scan_yaml_files(
+        shell: &Shell,
+        dir: &Path,
+        ecosystem_ports: &mut EcosystemPorts,
+    ) -> Result<()> {
+        for path in shell.read_dir(dir)? {
+            if !path.is_file() {
+                continue;
+            }
+            if let Some(extension) = path.extension() {
+                if extension == "yaml" || extension == "yml" {
+                    Self::process_yaml_file(shell, &path, ecosystem_ports)
+                        .context(format!("Error processing YAML file {:?}", path))?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn process_yaml_file(
+        shell: &Shell,
+        file_path: &Path,
+        ecosystem_ports: &mut EcosystemPorts,
+    ) -> Result<()> {
+        let contents = shell.read_file(file_path)?;
+        let value: Value = serde_yaml::from_str(&contents)?;
+        Self::traverse_yaml(&value, "", file_path, ecosystem_ports);
+        Ok(())
+    }
+
+    fn traverse_yaml(
+        value: &Value,
+        path: &str,
+        file_path: &Path,
+        ecosystem_ports: &mut EcosystemPorts,
+    ) {
+        match value {
+            Value::Mapping(map) => {
+                for (key, val) in map {
+                    let new_path = if path.is_empty() {
+                        key.as_str().unwrap_or_default().to_string()
+                    } else {
+                        format!("{}:{}", path, key.as_str().unwrap_or_default())
+                    };
+
+                    if key.as_str() == Some("ports") {
+                        if let Value::Sequence(ports) = val {
+                            for port_entry in ports {
+                                if let Some(port_str) = port_entry.as_str() {
+                                    let parts: Vec<&str> = port_str.split(':').collect();
+
+                                    if parts.len() > 1 {
+                                        if let Some(host_port_str) = parts.get(parts.len() - 2) {
+                                            if let Ok(port) = host_port_str.parse::<u16>() {
+                                                let description = format!(
+                                                    "[{}] {}",
+                                                    file_path.display(),
+                                                    new_path
+                                                );
+                                                ecosystem_ports.add_port_info(port, description);
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        }
```

Comment:
i'd really appreciate if you split this logic to smaller functions and add tests for them